File: Leetcode 3501-3600/3568/3568.java
// Leetcode 3568: Minimum Moves to Clean the Classroom
// https://leetcode.com/problems/minimum-moves-to-clean-the-classroom/
// Solved on 8th of November, 2025
import java.util.ArrayDeque;

class Solution {
    private static class State {
        int r;
        int c;
        int mask;
        int energy;
        int moves;
        State(int r, int c, int mask, int energy, int moves) {
            this.r = r;
            this.c = c;
            this.mask = mask;
            this.energy = energy;
            this.moves = moves;
        }
    }
    /**
     * Calculates the minimum number of moves required to clean all litter in the classroom.
     *
     * @param classroom A 2D array of characters representing the classroom grid. 'S' is the start, 'L' is litter, 'R' is a recharge station, 'X' is an obstacle.
     * @param energy The initial and maximum energy of the cleaner.
     * @return The minimum number of moves to clean all litter, or -1 if it's not possible.
     */
    public int minMoves(String[] classroom, int energy) {
        int rows = classroom.length;
        int cols = classroom[0].length();
        char[][] grid = new char[rows][cols];
        int startR = -1, startC = -1;
        int[][] litterIndex = new int[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                litterIndex[i][j] = -1;
            }
        }
        int litterCount = 0;
        for (int i = 0; i < rows; i++) {
            String s = classroom[i];
            for (int j = 0; j < cols; j++) {
                char ch = s.charAt(j);
                grid[i][j] = ch;
                if (ch == 'S') {
                    startR = i;
                    startC = j;
                } else if (ch == 'L') {
                    litterIndex[i][j] = litterCount++;
                }
            }
        }
        if (litterCount == 0) {
            return 0;
        }
        int allMask = (1 << litterCount) - 1;
        int maxEnergy = energy;
        int[][][] bestEnergy = new int[rows][cols][1 << litterCount];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                for (int k = 0; k < (1 << litterCount); k++) {
                    bestEnergy[i][j][k] = -1;
                }
            }
        }
        ArrayDeque<State> queue = new ArrayDeque<>();
        bestEnergy[startR][startC][0] = energy;
        queue.add(new State(startR, startC, 0, energy, 0));
        int[] dr = {-1,1,0,0};
        int[] dc = {0,0,-1,1};
        while (!queue.isEmpty()) {
            State cur = queue.poll();
            if (cur.mask == allMask) {
                return cur.moves;
            }
            if (cur.energy == 0) {
                continue;
            }
            for (int d = 0; d < 4; d++) {
                int nr = cur.r + dr[d];
                int nc = cur.c + dc[d];
                if (nr < 0 || nr >= rows || nc < 0 || nc >= cols) {
                    continue;
                }
                char cell = grid[nr][nc];
                if (cell == 'X') {
                    continue;
                }
                int nextEnergy = cur.energy - 1;
                if (cell == 'R') {
                    nextEnergy = maxEnergy;
                }
                int nextMask = cur.mask;
                if (cell == 'L') {
                    int idx = litterIndex[nr][nc];
                    if (idx >= 0) nextMask = cur.mask | (1 << idx);
                }
                if (bestEnergy[nr][nc][nextMask] < nextEnergy) {
                    bestEnergy[nr][nc][nextMask] = nextEnergy;
                    queue.add(new State(nr, nc, nextMask, nextEnergy, cur.moves + 1));
                }
            }
        }
        return -1;
    }
}

File: Leetcode 3501-3600/3568/3568.py
# Leetcode 3568: Minimum Moves to Clean the Classroom
# https://leetcode.com/problems/minimum-moves-to-clean-the-classroom/
# Solved on 8th of November, 2025
import collections


class Solution:
    def minMoves(self, classroom: list[str], energy: int) -> int:
        """
        Calculates the minimum number of moves required to clean all litter in the classroom.

        Args:
            classroom: A list of strings representing the classroom grid.
            energy: The initial energy of the robot.

        Returns:
            The minimum number of moves to clean all litter, or -1 if it's impossible.
        """
        numRows = len(classroom)
        numCols = len(classroom[0])
        litterCoords = {}
        startRow, startCol = -1, -1
        litterCount = 0

        for r in range(numRows):
            for c in range(numCols):
                if classroom[r][c] == 'S':
                    startRow, startCol = r, c
                elif classroom[r][c] == 'L':
                    litterCoords[(r, c)] = litterCount
                    litterCount += 1

        if litterCount == 0:
            return 0

        targetMask = (1 << litterCount) - 1

        queue = collections.deque()
        queue.append((startRow, startCol, 0, energy))

        visited = {}
        visited[(startRow, startCol, 0)] = energy

        currMoves = 0

        while queue:
            levelSize = len(queue)
            for _ in range(levelSize):
                r, c, currMask, currEnergy = queue.popleft()

                if currMask == targetMask:
                    return currMoves

                energyOnMove = -1
                if currEnergy > 0:
                    energyOnMove = currEnergy
                elif currEnergy == 0 and classroom[r][c] == 'R':
                    energyOnMove = energy

                if energyOnMove == -1:
                    continue

                for dr, dc in [(0, 1), (0, -1), (1, 0), (-1, 0)]:
                    nextRow, nextCol = r + dr, c + dc

                    if 0 <= nextRow < numRows and 0 <= nextCol < numCols and classroom[nextRow][nextCol] != 'X':
                        nextEnergy = energyOnMove - 1
                        nextMask = currMask
                        cellType = classroom[nextRow][nextCol]

                        if cellType == 'R':
                            nextEnergy = energy
                        elif cellType == 'L':
                            litterIndex = litterCoords.get((nextRow, nextCol))
                            if litterIndex is not None:
                                nextMask = currMask | (1 << litterIndex)

                        nextState = (nextRow, nextCol, nextMask)
                        if nextState not in visited or visited[nextState] < nextEnergy:
                            visited[nextState] = nextEnergy
                            queue.append((nextRow, nextCol, nextMask, nextEnergy))

            currMoves += 1

        return -1